api/lib/domain/usecases/update-target-profile.js

```diff
@@ -1,25 +1,23 @@
-import { validate } from '../validators/target-profile/base-validation.js';
+import { NotFoundError } from '../../domain/errors.js';
```

Comment:
j'ai un avis de plus en plus tranché sur la pertinence de l'existence des usecases.
On a un cas parfait où il n'a pas vraiment de responsabilités, de pouvoir de décision puisque tout a été déplacé dans le modèle.
On pourrait aussi bien déplacer tout ce code dans le controller que personnellement ca ne me ferait ni chaud ni froid, et sans casser l'archi hexagonale.
Le truc qui met le clou final dans le cercueil pour moi c'est qu'on se retrouve à écrire un fichier de test unitaire de ce usecase, qui est assez peu utile et surtout très difficilement maintenable car dès qu'on change un détail d'implémentation le test se transforme en guirlande de Noël. Il mock tout dans tous les sens, il est bien trop proche de l'implémentation

Reply:
Je remets la définition des usecases dans le livre "Clean Architecture" pour qu'on ait la même base dans la discussion

> Use cases contain the rules that specify how and when the Critical Business Rules within the Entities are invoked. Use cases control the dance of the Entities.
Notice also that the use case does not describe the user interface other than to informally specify the data coming in from that interface, and the data going back out through that interface. From the use case, it is impossible to tell whether the application is delivered on the web, or on a thick client, or on a console, or is a pure service.
This is very important.

Il y a certain usecases où je suis d'accord avec toi @laura-bergoens sur le fait qu'ils sont des passes plat, comme celui là par exemple : 

```js
const acceptPixCertifTermsOfService = function ({ userId, userRepository }) {
  return userRepository.updatePixCertifTermsOfServiceAcceptedToTrue(userId);
};

export { acceptPixCertifTermsOfService };
```

Dans celui-là, effectivement le usecase ne fait que transmettre au repository les informations qu'il a en entrée et ne manipule donc aucun objet du domain. Il ne répond donc pas à la définition d'un usecase.

Par contre on pourrait le refactorer de la manière suivante pour qu'il en soit plus proche : 

```js
const acceptPixCertifTermsOfService = function ({ userId, userRepository }) {
  const user = userRepository.get(userId);

  user.acceptTermsOfService({ certif: true });

  return userRepository.update(user)
};

export { acceptPixCertifTermsOfService };
```

Et donc pour moi le usecase updateTargetProfile n'est pas juste un passe plat : 
- Il prend en entrée des données dans un certain format
- Il va chercher le model du domain concerné
- Appelle une méthode du model qui encapsule la logique métier
- Persiste le model 

Pour moi, si on appelle directement les repository dans le controller et qu'on manipule les models à ce niveau là on casse le découplage.

> Use cases do not describe how the system appears to the user. Instead, they describe the application-specific rules that govern the interaction between the users and the Entities. How the data gets in and out of the system is irrelevant to the use cases.

Dans un commentaire plus haut tu parles des différents type de validation https://github.com/1024pix/pix/pull/8842#discussion_r1593172112. Si tout le code est dans le controller qu'est ce qui nous empêche de céder à la facilité de tout mettre dans la configuration du endpoint ?

> Use cases expect input data, and they produce output data. However, a well-formed use case object should have no inkling about the way that data is communicated to the user, or to any other component. We certainly don’t want the code within the use case class to know about HTML or SQL!

Les controllers sont pour moi couplé à notre contexte d'utilisation de l'API (sur le web) et les libraries ou patterns qu'on utilise (ex : Hapi, json-api, ...). Ce qui fait que si demain on décide d'en changer (exemple au hasard GraphQL ce que permet EmberData) on va devoir potentiellement toucher à du code qui est sensé être dans le domain si on a pas été en mesure de bien décorrelé les deux.

> Le truc qui met le clou final dans le cercueil pour moi c'est qu'on se retrouve à écrire un fichier de test unitaire de ce usecase, qui est assez peu utile et surtout très difficilement maintenable car dès qu'on change un détail d'implémentation le test se transforme en guirlande de Noël. Il mock tout dans tous les sens, il est bien trop proche de l'implémentation 

Cela me semble normal puisque le usecase est un chef d'orchestre son seul rôle est de s'assurer qu'un ensemble d'actions a été éxécuté dans le bon ordre. Pour moi un bon usecase ne doit pas avoir de logique. Car si c'est le cas c'est que vraisemblablement c'est du métier et donc a plus sa place dans un model.

Un dernier point qui m'embête avec le fait de mettre directement l'appel aux repository et la manipulation des models dans les controllers c'est faire reposer sur les épaules des développeurs la décision de ce qui merite un fichier de usecase ou pas. C'est une heuristique compliquée et très subjective (la réponse variera d'un dev à un autre). 

Quand quelqu'un arrive chez Pix, il doit se familiariser avec la Clean Architecture et DDD. Pour les devs juniors cela n'est pas évident du tout. On a régulièrement des remarques sur les différences de faire à certains endroits. Si l'architecture varie cela va renforcer ce sentiment et la réponse aux questions a apporté ne sera pas évidente de mon point de vue.

Reply:
**Précision suite à une discussion Slack** Ce thread est plus un échange d'idée qu'une vraie demande de changement dans le cadre de cette PR.